### DISCUSSION TITLE
get logs from `crashpad_handler` on linux

### DISCUSSION THREAD

**Discussion by kristjanvalur** [getsentry/sentry-native#1010] [Q&A] [answered]
So, we are running a containerized service using sentry-native and crashpad.
Once the main process quits, we look for a "crashpad_handler" process, and wait for it to exit before exiting the container.

This works, but it would still, for logging purposes, be useful to **verify** that crashpad had actually successfully uploaded the report to sentry.  Is there a way to extract any logging from crashpad, after run, to include in the container log?  Could we instruct it, e.g. via an environment variable, to log to a specific file?

I took a cursory look at crashpad, and I couldn't see that anything useful were being logged at all, and in fact, it looked to me that on linux the log handlers were in fact no-ops.  

Any thoughts?

**Comment by supervacuus** [comment, marked as answer]
> Could we instruct it, e.g. via an environment variable, to log to a specific file?

Not without changing its logging implementation. The logging configuration is not exposed to the client but rather hard-coded depending on the build configuration (mainly the target OS). So, on Linux, it is set to log to `stderr` and the system log (which would be helpful in a container setup), but it seems to be the only supported operating system with no implementation for that option :-)

> I took a cursory look at crashpad, and I couldn't see that anything useful were being logged at all, and in fact, it looked to me that on linux the log handlers were in fact no-ops.

That is true. `crashpad` doesn't log helpful state information. It mostly follows the premise of logging on errors or fatal errors since Google doesn't use C++ exceptions, and fatal logs, in many cases, are their primary mechanism to exit on error. 

Logging on Linux should work, though. `crashpad_handler` stderr logs route back to the client, but that might not help much since the client is often already terminated at the point of a successful upload.

> This works, but it would still, for logging purposes, be useful to verify that crashpad had actually successfully uploaded the report to sentry.

While one can certainly approach this topic from a perspective of syslog-wrapping or starting the application (and processes it starts) as a `systemd` service, I think the most straightforward approach is to log the contents of the directory `<database-path>/completed` after waiting for the `crashpad_handler` to quit, because it will move minidumps there only when uploading was successful.

**Reply by kristjanvalur** [reply]
This sounds like an excellent suggestion.  I'll have my container script list the contents of this folder.  Now, if I could only remember what I set it to be :)
